pyiceberg/table/__init__.py

```diff
@@ -502,6 +503,73 @@ def append(self, df: pa.Table, snapshot_properties: Dict[str, str] = EMPTY_DICT)
                 for data_file in data_files:
                     append_files.append_data_file(data_file)
 
+    def _build_partition_predicate(self, partition_records: List[Record]) -> BooleanExpression:
```

Comment:
Can you add two tests where:

- We start with an unpartitioned table, write some data, and then evolve to a partition.
- Start with a Monthly partitioned table, insert data for a few days, convert the partition to a daily partition, and dynamically overwrite a single day.

Reply:
Hi Fokko, thanks for the detailed review! I am working on fixing them. 

For this comment, I would like to check your opinions on what should be the behavior in the first scenario, i.e. a dynamic overwrite's impact on files that do not have partitions? Cus currently I alerts the user the dynamic overwrite does not makes sense if current table is unpartitioned. I think I could change to a no-op operation. If so, it would leave us in a slightly weird situation where an unpartitioned table would have all the data overwritten if it is normal overwrite but it would end up as a no-op if it is dynamic overwrite, does that sound correct to you?

Reply:
Thanks for the thorough review! So if the table is firstly unpartitioned and then evolved into a partition. I think the expected bahavior is that dynamic overwrite will also delete (potentially through overwrite) data in the unpartitioned files?